File: app/src/main/java/com/andalus/abomed7at55/mn_edek_a7la/Adapters/CategoriesAdapter.java
package com.andalus.abomed7at55.mn_edek_a7la.Adapters;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.andalus.abomed7at55.mn_edek_a7la.Objects.FoodCategory;
import com.andalus.abomed7at55.mn_edek_a7la.R;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class CategoriesAdapter extends RecyclerView.Adapter<CategoriesAdapter.CategoryHolder>{

    private List<FoodCategory> mData;

    public CategoriesAdapter(List<FoodCategory> data){
        mData = data;
    }

    @NonNull
    @Override
    public CategoryHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view = inflater.inflate(R.layout.category_item,parent,false);
        return new CategoryHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull CategoryHolder holder, int position) {
        holder.ivCategoryIcon.setImageResource(mData.get(position).getIconId());
        holder.tvCategoryTitle.setText(mData.get(position).getTitle());
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }

    class CategoryHolder extends RecyclerView.ViewHolder{

        @BindView(R.id.iv_category_icon)
        ImageView ivCategoryIcon;
        @BindView(R.id.tv_category_title)
        TextView tvCategoryTitle;

        CategoryHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this,itemView);
        }
    }
}

File: app/src/main/java/com/andalus/abomed7at55/mn_edek_a7la/adapters/CategoriesAdapter.kt
package com.andalus.abomed7at55.mn_edek_a7la.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.LinearSnapHelper
import androidx.recyclerview.widget.RecyclerView
import com.andalus.abomed7at55.mn_edek_a7la.R
import com.andalus.abomed7at55.mn_edek_a7la.model.Category
import kotlinx.android.synthetic.main.item_category.view.*
import java.io.Serializable

class CategoriesAdapter(
        private val onCategoryClicked: (category: String) -> Unit = {},
        var data: List<Category> = listOf(),
        private val onRecipeClicked: (id: Int) -> Unit = {},
        private val onOptionsClicked: (id: Int, optionsButton: View) -> Unit
) : RecyclerView.Adapter<CategoriesAdapter.CategoryHolder>(), Serializable {

    private lateinit var context: Context

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CategoryHolder {
        context = parent.context
        val view = LayoutInflater.from(context).inflate(R.layout.item_category, parent, false)
        return CategoryHolder(view)
    }

    override fun onBindViewHolder(holder: CategoryHolder, position: Int) {
        if (data[position].title == context.getString(R.string.recent)) {
            holder.btnViewAll.visibility = View.INVISIBLE
        }

        holder.tvCategoryName.text = data[position].title

        data[position].recipesAdapter.onClick = {
            onRecipeClicked.invoke(it)
        }

        data[position].recipesAdapter.onOptionsClicked = { id, optionsButton ->
            onOptionsClicked.invoke(id, optionsButton)
        }

        holder.rvRecipes.adapter = data[position].recipesAdapter

    }

    override fun getItemCount(): Int {
        return data.size
    }

    inner class CategoryHolder(view: View) : RecyclerView.ViewHolder(view) {
        val tvCategoryName: TextView = itemView.tvCategoryName
        val rvRecipes: RecyclerView = itemView.rvRecipes
        val btnViewAll: Button = itemView.btnViewAll

        init {
            rvRecipes.layoutManager = LinearLayoutManager(context, RecyclerView.HORIZONTAL, true)
            rvRecipes.onFlingListener = null
            val snapHelper = LinearSnapHelper()
            snapHelper.attachToRecyclerView(rvRecipes)
            btnViewAll.setOnClickListener {
                onCategoryClicked.invoke(data[adapterPosition].tag)
            }
        }
    }
}